java.nio.charset.CoderMalfunctionError: IllegalArgumentException: newPosition > limit: (2 > 1) & NPE Cannot invoke "java.nio.Buffer.arrayOffset()" because "dst" is null

Failure link
------------

From an internal build `job/Test_openjdk17_j9_extended.openjdk_x86-64_mac/44/`(`mac11x64rt06 `):
```
openjdk version "17.0.2" 2022-01-18
IBM Semeru Runtime Certified Edition 17.0.2.0 (build 17.0.2+8)
Eclipse OpenJ9 VM 17.0.2.0 (build openj9-0.30.1, JRE 17 Mac OS X amd64-64-Bit Compressed References 20220209_95 (JIT enabled, AOT enabled)
OpenJ9   - 9dccbe076
OMR      - dac962a28
JCL      - 316ae6c0611 based on jdk-17.0.2+8)
```
[Rerun in Grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/parambuild/?SDK_RESOURCE=upstream&amp;TARGET=jdk_nio_0&amp;TEST_FLAG=&amp;UPSTREAM_TEST_JOB_NAME=&amp;DOCKER_REQUIRED=false&amp;ACTIVE_NODE_TIMEOUT=0&amp;VENDOR_TEST_DIRS=&amp;EXTRA_DOCKER_ARGS=&amp;TKG_OWNER_BRANCH=AdoptOpenJDK%3Amaster&amp;OPENJ9_SYSTEMTEST_OWNER_BRANCH=eclipse%3Amaster&amp;PLATFORM=x86-64_mac&amp;GENERATE_JOBS=true&amp;KEEP_REPORTDIR=true&amp;PERSONAL_BUILD=false&amp;ADOPTOPENJDK_REPO=https%3A%2F%2Fgithub.com%2Fadoptium%2Faqa-tests.git&amp;LABEL=&amp;EXTRA_OPTIONS=&amp;CUSTOMIZED_SDK_URL=+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk17u%2Fjdk17u-mac-x64-openj9-IBM%2F95%2Fibm-semeru-certified-jdk_x64_mac_17.0.2.0.tar.gz+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk17u%2Fjdk17u-mac-x64-openj9-IBM%2F95%2Fibm-semeru-certified-debugimage_x64_mac_17.0.2.0.tar.gz+https%3A%2F%2Fna.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2Fbuild-scripts%2Fjobs%2Fjdk17u%2Fjdk17u-mac-x64-openj9-IBM%2F95%2Fibm-semeru-certified-testimage_x64_mac_17.0.2.0.tar.gz&amp;BUILD_IDENTIFIER=&amp;ADOPTOPENJDK_BRANCH=master&amp;LIGHT_WEIGHT_CHECKOUT=true&amp;NON_AQA_TEST_REPOS=&amp;ARTIFACTORY_SERVER=na.artifactory.swg-devops&amp;KEEP_WORKSPACE=false&amp;USER_CREDENTIALS_ID=&amp;JDK_VERSION=17&amp;ITERATIONS=1&amp;VENDOR_TEST_REPOS=&amp;JDK_REPO=git%40github.com%3Aibmruntimes%2Fopenj9-openjdk-jdk17&amp;RELEASE_TAG=v0.30.0-release&amp;OPENJ9_BRANCH=v0.30.0-release&amp;OPENJ9_SHA=&amp;JCK_GIT_REPO=&amp;VENDOR_TEST_BRANCHES=&amp;OPENJ9_REPO=https%3A%2F%2Fgithub.com%2Feclipse-openj9%2Fopenj9.git&amp;UPSTREAM_JOB_NAME=&amp;CLOUD_PROVIDER=&amp;PLATFORM_AND_MACHINE=&amp;CUSTOM_TARGET=&amp;VENDOR_TEST_SHAS=&amp;JDK_BRANCH=v0.30.0-release&amp;LABEL_ADDITION=&amp;ARTIFACTORY_REPO=&amp;ARTIFACTORY_ROOT_DIR=&amp;UPSTREAM_TEST_JOB_NUMBER=&amp;DOCKERIMAGE_TAG=&amp;JDK_IMPL=openj9&amp;TEST_TIME=&amp;SSH_AGENT_CREDENTIAL=83181e25-eea4-4f55-8b3e-e79615733226&amp;AUTO_DETECT=true&amp;TKG_SHA=&amp;ADOPTOPENJDK_SYSTEMTEST_OWNER_BRANCH=adoptium%3Amaster&amp;CUSTOMIZED_SDK_URL_CREDENTIAL_ID=4e18ffe7-b1b1-4272-9979-99769b68bcc2&amp;NUM_MACHINES=&amp;OPENJDK_SHA=&amp;TRSS_URL=http%3A%2F%2Ftrss1.fyre.ibm.com&amp;BUILD_LIST=openjdk&amp;UPSTREAM_JOB_NUMBER=&amp;STF_OWNER_BRANCH=adoptium%3Amaster&amp;TIME_LIMIT=20&amp;JVM_OPTIONS=&amp;PARALLEL=None) - Change TARGET to run only the failed test targets.

Optional info
-------------

Failure output (captured from console output)
---------------------------------------------
```
[2022-02-09T17:04:48.332Z] variation: Mode150
[2022-02-09T17:04:48.332Z] JVM_OPTIONS:  -XX:+UseCompressedOops 

[2022-02-09T17:11:36.665Z] TEST: sun/nio/cs/FindDecoderBugs.java

[2022-02-09T17:11:36.667Z] STDERR:
[2022-02-09T17:11:36.667Z] java.nio.charset.CoderMalfunctionError: java.lang.IllegalArgumentException: newPosition > limit: (2 > 1)
[2022-02-09T17:11:36.667Z] 	at java.base/java.nio.charset.CharsetDecoder.decode(CharsetDecoder.java:587)
[2022-02-09T17:11:36.667Z] 	at FindDecoderBugs$CharsetTester.recode(FindDecoderBugs.java:225)
[2022-02-09T17:11:36.667Z] 	at FindDecoderBugs$CharsetTester.recode2(FindDecoderBugs.java:256)
[2022-02-09T17:11:36.667Z] 	at FindDecoderBugs$CharsetTester.test(FindDecoderBugs.java:279)
[2022-02-09T17:11:36.667Z] 	at FindDecoderBugs$CharsetTester.testRandomly(FindDecoderBugs.java:308)
[2022-02-09T17:11:36.667Z] 	at FindDecoderBugs$CharsetTester.testPrefix(FindDecoderBugs.java:323)
[2022-02-09T17:11:36.667Z] 	at FindDecoderBugs.testCharset(FindDecoderBugs.java:399)
[2022-02-09T17:11:36.667Z] 	at FindDecoderBugs.realMain(FindDecoderBugs.java:420)
[2022-02-09T17:11:36.667Z] 	at FindDecoderBugs.main(FindDecoderBugs.java:446)
[2022-02-09T17:11:36.667Z] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[2022-02-09T17:11:36.667Z] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
[2022-02-09T17:11:36.667Z] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2022-02-09T17:11:36.667Z] 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
[2022-02-09T17:11:36.667Z] 	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:312)
[2022-02-09T17:11:36.667Z] 	at java.base/java.lang.Thread.run(Thread.java:884)
[2022-02-09T17:11:36.667Z] Caused by: java.lang.IllegalArgumentException: newPosition > limit: (2 > 1)
[2022-02-09T17:11:36.667Z] 	at java.base/java.nio.Buffer.createPositionException(Buffer.java:341)
[2022-02-09T17:11:36.667Z] 	at java.base/java.nio.Buffer.position(Buffer.java:316)
[2022-02-09T17:11:36.667Z] 	at java.base/java.nio.CharBuffer.position(CharBuffer.java:1516)
[2022-02-09T17:11:36.667Z] 	at java.base/sun.nio.cs.DoubleByte$Decoder.decodeArrayLoop(DoubleByte.java:199)
[2022-02-09T17:11:36.667Z] 	at java.base/sun.nio.cs.DoubleByte$Decoder.decodeLoop(DoubleByte.java:233)
[2022-02-09T17:11:36.667Z] 	at java.base/java.nio.charset.CharsetDecoder.decode(CharsetDecoder.java:585)
[2022-02-09T17:11:36.667Z] 	... 14 more

[2022-02-09T17:11:36.672Z] java.nio.charset.CoderMalfunctionError: java.lang.IllegalArgumentException: newPosition > limit: (3 > 2)
[2022-02-09T17:11:36.672Z] 	at java.base/java.nio.charset.CharsetDecoder.decode(CharsetDecoder.java:587)
[2022-02-09T17:11:36.672Z] 	at FindDecoderBugs$CharsetTester.recode(FindDecoderBugs.java:225)
[2022-02-09T17:11:36.672Z] 	at FindDecoderBugs$CharsetTester.recode2(FindDecoderBugs.java:256)
[2022-02-09T17:11:36.672Z] 	at FindDecoderBugs$CharsetTester.test(FindDecoderBugs.java:279)
[2022-02-09T17:11:36.672Z] 	at FindDecoderBugs$CharsetTester.testRandomly(FindDecoderBugs.java:308)
[2022-02-09T17:11:36.672Z] 	at FindDecoderBugs$CharsetTester.testPrefix(FindDecoderBugs.java:324)
[2022-02-09T17:11:36.672Z] 	at FindDecoderBugs.testCharset(FindDecoderBugs.java:399)
[2022-02-09T17:11:36.672Z] 	at FindDecoderBugs.realMain(FindDecoderBugs.java:420)
[2022-02-09T17:11:36.672Z] 	at FindDecoderBugs.main(FindDecoderBugs.java:446)
[2022-02-09T17:11:36.672Z] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[2022-02-09T17:11:36.672Z] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
[2022-02-09T17:11:36.672Z] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2022-02-09T17:11:36.672Z] 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
[2022-02-09T17:11:36.672Z] 	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:312)
[2022-02-09T17:11:36.672Z] 	at java.base/java.lang.Thread.run(Thread.java:884)
[2022-02-09T17:11:36.672Z] Caused by: java.lang.IllegalArgumentException: newPosition > limit: (3 > 2)
[2022-02-09T17:11:36.672Z] 	at java.base/java.nio.Buffer.createPositionException(Buffer.java:341)
[2022-02-09T17:11:36.672Z] 	at java.base/java.nio.Buffer.position(Buffer.java:316)
[2022-02-09T17:11:36.672Z] 	at java.base/java.nio.CharBuffer.position(CharBuffer.java:1516)
[2022-02-09T17:11:36.672Z] 	at java.base/sun.nio.cs.DoubleByte$Decoder.decodeArrayLoop(DoubleByte.java:199)
[2022-02-09T17:11:36.672Z] 	at java.base/sun.nio.cs.DoubleByte$Decoder.decodeLoop(DoubleByte.java:233)
[2022-02-09T17:11:36.672Z] 	at java.base/java.nio.charset.CharsetDecoder.decode(CharsetDecoder.java:585)
[2022-02-09T17:11:36.672Z] 	... 14 more
[2022-02-09T17:11:36.672Z] java.lang.AssertionError: Some tests failed
[2022-02-09T17:11:36.672Z] 	at FindDecoderBugs.main(FindDecoderBugs.java:448)
[2022-02-09T17:11:36.672Z] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[2022-02-09T17:11:36.672Z] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
[2022-02-09T17:11:36.672Z] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[2022-02-09T17:11:36.672Z] 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
[2022-02-09T17:11:36.672Z] 	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:312)
[2022-02-09T17:11:36.672Z] 	at java.base/java.lang.Thread.run(Thread.java:884)
[2022-02-09T17:11:36.672Z] 
[2022-02-09T17:11:36.672Z] JavaTest Message: Test threw exception: java.lang.AssertionError
[2022-02-09T17:11:36.672Z] JavaTest Message: shutting down test

[2022-02-09T17:11:36.673Z] 
[2022-02-09T17:11:36.673Z] Passed = 524916382, failed = 6081
[2022-02-09T17:11:36.673Z] 
[2022-02-09T17:11:36.673Z] 
[2022-02-09T17:11:36.673Z] TEST RESULT: Failed. Execution failed: `main' threw exception: java.lang.AssertionError: Some tests failed
[2022-02-09T17:11:36.673Z] --------------------------------------------------
[2022-02-09T17:16:48.534Z] Test results: passed: 436; failed: 1
[2022-02-09T17:16:58.094Z] Report written to /Users/jenkins/workspace/Test_openjdk17_j9_extended.openjdk_x86-64_mac/jvmtest/openjdk/report/html/report.html
[2022-02-09T17:16:58.094Z] Results written to /Users/jenkins/workspace/Test_openjdk17_j9_extended.openjdk_x86-64_mac/aqa-tests/TKG/output_16444255309754/jdk_nio_0/work
[2022-02-09T17:16:58.094Z] Error: Some tests failed or other problems occurred.
[2022-02-09T17:16:58.094Z] 
[2022-02-09T17:16:58.094Z] jdk_nio_0_FAILED
```

This appears https://github.com/eclipse-openj9/openj9/issues/13983 though this build contains the fix https://github.com/eclipse-openj9/openj9/pull/14294

50x grinder (`jdk_nio_0`) - /job/Grinder/21548/

`job/Grinder_iteration_3/692/consoleFull`
```
13:42:17  variation: Mode150
13:42:17  JVM_OPTIONS:  -XX:+UseCompressedOops 

13:49:07  TEST: sun/nio/cs/TestStringCoding.java

13:49:07  STDERR:
13:49:07  java.nio.charset.CoderMalfunctionError: java.lang.NullPointerException: Cannot store to char array because "da" is null
13:49:07  	at java.base/java.nio.charset.CharsetDecoder.decode(CharsetDecoder.java:587)
13:49:07  	at java.base/java.nio.charset.CharsetDecoder.decode(CharsetDecoder.java:806)
13:49:07  	at TestStringCoding.testNewString(TestStringCoding.java:180)
13:49:07  	at TestStringCoding.main(TestStringCoding.java:88)
13:49:07  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
13:49:07  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
13:49:07  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
13:49:07  	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
13:49:07  	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
13:49:07  	at java.base/java.lang.Thread.run(Thread.java:884)
13:49:07  Caused by: java.lang.NullPointerException: Cannot store to char array because "da" is null
13:49:07  	at java.base/sun.nio.cs.DoubleByte$Decoder.decodeArrayLoop(DoubleByte.java:192)
13:49:07  	at java.base/sun.nio.cs.DoubleByte$Decoder.decodeLoop(DoubleByte.java:233)
13:49:07  	at java.base/java.nio.charset.CharsetDecoder.decode(CharsetDecoder.java:585)
13:49:07  	... 9 more
```

500x grinder (`sun/nio/cs/FindDecoderBugs.java`) - `view/Test_grinder/job/Grinder/21554/` - all passed

I tried to be specific according to your recommendation and used `-Xjit:exclude={*CharsetDecoder.decode*}`, because there are man `decode*` methods in both JRE and third-party libraries. Several builds passed. But [this one](https://github.com/dev-aspectj/aspectj-maven-plugin/runs/5959366388?check_suite_focus=true#step:6:11053) now shows a slightly different stack trace:

```text
Caused by: java.lang.IllegalArgumentException: newPosition > limit: (0 > -1)
	at java.base/java.nio.Buffer.createPositionException(Buffer.java:341)
	at java.base/java.nio.Buffer.position(Buffer.java:316)
	at java.base/java.nio.CharBuffer.position(CharBuffer.java:1516)
	at java.base/sun.nio.cs.ISO_8859_1$Decoder.decodeArrayLoop(ISO_8859_1.java:93)
	at java.base/sun.nio.cs.ISO_8859_1$Decoder.decodeLoop(ISO_8859_1.java:122)
	at java.base/java.nio.charset.CharsetDecoder.decode(CharsetDecoder.java:585)
```

That one I am seeing for the very first time. Before, it was always the NPE. So the problem does not seem to be (only) in `CharsetDecoder.decode`. I am going to try and exclude more in the next iteration.

--- a/runtime/compiler/x/codegen/J9TreeEvaluator.cpp
+++ b/runtime/compiler/x/codegen/J9TreeEvaluator.cpp
@@ -11905,8 +11905,8 @@ J9::X86::TreeEvaluator::inlineStringLatin1Inflate(TR::Node *node, TR::CodeGenera
    TR::Register *zeroReg = cg->allocateRegister(TR_VRF);
    TR::Register *scratchReg = cg->allocateRegister(TR_GPR);
 
-   int depCount = 9;
-   TR::RegisterDependencyConditions *deps = generateRegisterDependencyConditions((uint8_t)0, depCount, cg);
+   int maxDepCount = 10;
+   TR::RegisterDependencyConditions *deps = generateRegisterDependencyConditions(0, maxDepCount, cg);
    deps->addPostCondition(xmmHighReg, TR::RealRegister::NoReg, cg);
    deps->addPostCondition(xmmLowReg, TR::RealRegister::NoReg, cg);
    deps->addPostCondition(zeroReg, TR::RealRegister::NoReg, cg);
@@ -11916,7 +11916,6 @@ J9::X86::TreeEvaluator::inlineStringLatin1Inflate(TR::Node *node, TR::CodeGenera
    deps->addPostCondition(scratchReg, TR::RealRegister::eax, cg);
    deps->addPostCondition(srcOffsetReg, TR::RealRegister::ecx, cg);
    deps->addPostCondition(destOffsetReg, TR::RealRegister::edx, cg);
-   deps->stopAddingConditions();
 
    TR::LabelSymbol *doneLabel = generateLabelSymbol(cg);
    TR::LabelSymbol *copyResidueLabel = generateLabelSymbol(cg);
@@ -12004,7 +12003,15 @@ J9::X86::TreeEvaluator::inlineStringLatin1Inflate(TR::Node *node, TR::CodeGenera
 
    bool is64bit = cg->comp()->target().is64Bit();
    // calculate address to jump too
-   generateRegMemInstruction(TR::InstOpCode::LEARegMem(is64bit), node, scratchReg, generateX86MemoryReference(copyResidueLabel, cg), cg);
+
+   TR::MemoryReference *residueLabelMR = generateX86MemoryReference(copyResidueLabel, cg);
+
+   if (cg->comp()->target().is64Bit() && residueLabelMR->getAddressRegister())
+      {
+      deps->addPostCondition(residueLabelMR->getAddressRegister(), TR::RealRegister::NoReg, cg);
+      }
+
+   generateRegMemInstruction(TR::InstOpCode::LEARegMem(is64bit), node, scratchReg, residueLabelMR, cg);
    generateRegRegInstruction(TR::InstOpCode::ADDRegReg(is64bit), node, lengthReg, scratchReg, cg);
 
    generateRegMemInstruction(TR::InstOpCode::LEARegMem(is64bit), node, srcOffsetReg, generateX86MemoryReference(srcBufferReg, srcOffsetReg, 0, 0, cg), cg);
@@ -12020,6 +12027,7 @@ J9::X86::TreeEvaluator::inlineStringLatin1Inflate(TR::Node *node, TR::CodeGenera
       generateMemRegInstruction(TR::InstOpCode::S2MemReg, node, generateX86MemoryReference(destOffsetReg, headerOffsetConst + 2 * (6 - i), cg), scratchReg, cg);
       }
 
+   deps->stopAddingConditions();
    generateLabelInstruction(TR::InstOpCode::label, node, doneLabel, deps, cg);
    doneLabel->setEndInternalControlFlow();
 

